### PR TITLE
fix: Change multiplexer scheduler to use two rails instead of three

### PR DESCRIPTION
**Pull request by arunkarthik-akkart**
Currently we  use 3 stripes for the below case:
(3*stripe_size) < payload size < (4*stripe_size)
After this commit we enforce the num_stripes always rounds down to the nearest lower power of 2.
We also set the default stripe size to be 128K instead of 64K since we see the most optimal results during testing.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/nccl_ofi_param.h [hunk 1]
```diff
@@ -249,7 +249,7 @@ OFI_NCCL_PARAM_INT(disable_dmabuf, "DISABLE_DMABUF", 0);
 /*
  * Messages sized larger than this threshold will be striped across multiple rails
  */
-OFI_NCCL_PARAM_UINT(min_stripe_size, "MIN_STRIPE_SIZE", (64 * 1024));
```
Comment (bwbarrett) [inline include/nccl_ofi_param.h:252]: Split this into its own commit, it deserves its own headline when you run `git log --oneline`.

File: src/nccl_ofi_scheduler.c [hunk 2]
```diff
@@ -42,6 +42,16 @@ static inline int set_schedule_by_threshold(nccl_net_ofi_threshold_scheduler_t *
 	/* Number of stripes is atleast 1 for zero-sized messages and at most equal to num of rails */
 	int num_stripes =
 		(int)NCCL_OFI_MAX(1, NCCL_OFI_MIN(NCCL_OFI_DIV_CEIL(size, scheduler->min_stripe_size), (unsigned)num_rails));
+
+	/* Ensure num_stripes is the nearest lower power of 2 */
```
Comment (bwbarrett) [inline src/nccl_ofi_scheduler.c:46]: This isn't what we agreed to implement?  Don't assume that we won't be stupid and do 6 NICs per GPU.

**Review by bwbarrett — CHANGES_REQUESTED** [review]

File: include/nccl_ofi_param.h (hunk 1 above)
Comment (bwbarrett) [reply include/nccl_ofi_param.h:252]: And you need a bit of why when you write the commit message for this change.

File: src/nccl_ofi_scheduler.c [hunk 3]
```diff
@@ -42,6 +42,16 @@ static inline int set_schedule_by_threshold(nccl_net_ofi_threshold_scheduler_t *
...
+	if (num_stripes > 1) {
+		int nearest_divisible_stripe_size = 1;
+		while (nearest_divisible_stripe_size * 2 <= num_stripes) {
+			nearest_divisible_stripe_size *= 2;
+		}
+		num_stripes = nearest_divisible_stripe_size;
+	}
+
 	if (OFI_UNLIKELY(num_rails == 0)) {
```
Comment (liralon) [inline src/nccl_ofi_scheduler.c:55]: Not related to this commit but spotted it during review: This check should be before setting `num_stripes`. Also, as long as `num_rails` is signed rather than unsigned, this should be `<= 0`.

File: src/nccl_ofi_scheduler.c [hunk 4]
```diff
@@ -42,6 +42,16 @@ static inline int set_schedule_by_threshold(nccl_net_ofi_threshold_scheduler_t *
 	/* Number of stripes is atleast 1 for zero-sized messages and at most equal to num of rails */
 	int num_stripes =
 		(int)NCCL_OFI_MAX(1, NCCL_OFI_MIN(NCCL_OFI_DIV_CEIL(size, scheduler->min_stripe_size), (unsigned)num_rails));
```
Comment (liralon) [inline src/nccl_ofi_scheduler.c:44]: Not related to this commit but spotted during review: It seems to me that `num_rails` should be `unsigned int` to begin with rather than signed.

File: include/nccl_ofi_param.h (hunk 1 above)
Comment (liralon) [reply include/nccl_ofi_param.h:252]: +1. As evident from `We also` in commit message, you are doing multiple logical changes in single commit rather than one.

File: src/nccl_ofi_scheduler.c (hunk 2 above)
Comment (liralon) [reply src/nccl_ofi_scheduler.c:46]: Right. I think you can have a simple loop as follows to find the closest number of rails to `num_stripes` that `num_rails` divides in:
```
for (int i = num_stripes; i >= 0; i--) {
	if ((num_rails % i) == 0) {
		num_stripes = i;
		break;
	}
}
```

Nit: I suggest to export above together with the calculation of existing `num_stripes` to a small utility function `get_num_stripes()`.

**Review by liralon — CHANGES_REQUESTED** [review]

File: include/nccl_ofi_param.h (hunk 1 above)
Comment (arunkarthik-akkart) [reply include/nccl_ofi_param.h:252]: Done. I have split the min_stripe_size from 64K to 128k to a separate commit.

File: src/nccl_ofi_scheduler.c (hunk 2 above)
Comment (arunkarthik-akkart) [reply src/nccl_ofi_scheduler.c]: Thanks for the detailed feedback. I have moved the implementation to get_num_stripes to a separate function as advised.

File: src/nccl_ofi_scheduler.c (hunk 3 above)
Comment (arunkarthik-akkart) [reply src/nccl_ofi_scheduler.c]: Done. I have changed the condition such that num_rails <= 0

File: src/nccl_ofi_scheduler.c (hunk 4 above)
Comment (arunkarthik-akkart) [reply src/nccl_ofi_scheduler.c]: Correct. There was some previous discussion in the original PR to make num_rails as unsigned int, but this will involve changing it everywhere in the code base.
I was thinking to make a separate PR for this change. But please advise if required, I can add it as an additional commit to this PR.

Comment (liralon) [reply src/nccl_ofi_scheduler.c]: I didn't meant to do this change as part of this PR. I implied that we should change that in a separate PR.

**Review by liralon — APPROVED** [review]
1. Add `Signed-off-by` tag to the end of your commit message.
2. Change `feat:` title prefix in commit changing `min_stripe_size` to be `param:` and `scheduler:` in the commit that changes how `num_stripes` is calculated.

The change itself looks fine to me now, so approving assuming you fix this prior to merge.

**Comment by xwangxin** [comment]
bot:aws:retest

**Comment by xwangxin** [comment]
bot:aws:retest

File: src/nccl_ofi_scheduler.c [hunk 5]
```diff
@@ -38,14 +72,14 @@ static inline int set_schedule_by_threshold(nccl_net_ofi_threshold_scheduler_t *
 					    nccl_net_ofi_schedule_t *schedule)
 {
 	int ret = 0;
+	int num_stripes = 0;
 
-	/* Number of stripes is atleast 1 for zero-sized messages and at most equal to num of rails */
-	int num_stripes =
-		(int)NCCL_OFI_MAX(1, NCCL_OFI_MIN(NCCL_OFI_DIV_CEIL(size, scheduler->min_stripe_size), (unsigned)num_rails));
-	if (OFI_UNLIKELY(num_rails == 0)) {
+	if (OFI_UNLIKELY(num_rails <= 0)) {
```
Comment (bwbarrett) [inline src/nccl_ofi_scheduler.c:77]: I missed in the previous commit that this was checking `num_rails` and not `num_stripes`.  This is a silly check and should just be an assert.

File: src/nccl_ofi_scheduler.c [hunk 6]
```diff
@@ -22,6 +22,40 @@ static inline size_t sizeof_schedule(int num_rails)
 		+ num_rails * sizeof(nccl_net_ofi_xfer_info_t);
 }
 
+/*
+ * @brief  This function calculates the optimal number of stripes
+ * for the payload size based on the min_stripe_size.
+ * It first determines the initial number of stripes based on the message size
+ * and a minimum stripe size, ensuring that the result is at least 1 and at
+ * most the number of available rails. The function then adjusts the number of
```
Comment (bwbarrett) [inline src/nccl_ofi_scheduler.c:30]: This does not match what the commit message says.  The code does not "modify the num_stripes to be rounded off to the nearest factor of the num_rails."

**Review by bwbarrett — CHANGES_REQUESTED** [review]

File: src/nccl_ofi_scheduler.c [hunk 7]
```diff
@@ -22,6 +22,40 @@ static inline size_t sizeof_schedule(int num_rails)
...
+ *
+ */
+static int get_num_stripes(nccl_net_ofi_threshold_scheduler_t *scheduler_p, size_t size, int num_rails)
+{
+	/* Number of stripes is atleast 1 for zero-sized messages and at most equal to num of rails */
+	int num_stripes =
+		(int)NCCL_OFI_MAX(1, NCCL_OFI_MIN(NCCL_OFI_DIV_CEIL(size, scheduler_p->min_stripe_size), (unsigned)num_rails));
+
+	/* Ensure num_stripes is the nearest lower factor of the number of rails */
+	for (int i = num_stripes; i >= 0; i--) {
```
Comment (bwbarrett) [inline src/nccl_ofi_scheduler.c:50]: You'll need a comment to explain, but this should be:

```
for (int i = num_stripes ; i > 1 ; i--)
```

if `num_stripes` is 1, then there's no need to run the check.  `num_rails % 1` will always be true, regardless of the value of `num_rails`.  And when `num_stripes` is 1, you're more likely in a latency sensitive regime, in which case the mod operation will increase latency.  So don't do things that increase latency and aren't necessary.

File: src/nccl_ofi_scheduler.c (hunk 5 above)
Comment (arunkarthik-akkart) [reply src/nccl_ofi_scheduler.c]: Done.

File: src/nccl_ofi_scheduler.c (hunk 6 above)
Comment (arunkarthik-akkart) [reply src/nccl_ofi_scheduler.c]: done. I have updated the commit message and the function description.

File: src/nccl_ofi_scheduler.c (hunk 7 above)
Comment (arunkarthik-akkart) [reply src/nccl_ofi_scheduler.c]: Thanks for pointing this and providing the detailed explanation. I have corrected this.

File: src/nccl_ofi_scheduler.c [hunk 8]
```diff
@@ -22,6 +22,43 @@ static inline size_t sizeof_schedule(int num_rails)
...
+static int get_num_stripes(nccl_net_ofi_threshold_scheduler_t *scheduler_p, size_t size, int num_rails)
+{
+	/* Number of stripes is atleast 1 for zero-sized messages and at most equal to num of rails */
+	int num_stripes =
+		(int)NCCL_OFI_MAX(1, NCCL_OFI_MIN(NCCL_OFI_DIV_CEIL(size, scheduler_p->min_stripe_size), (unsigned)num_rails));
+
+	/* Start the loop from num_stripes and skip 1, as num_rails % 1 is always true.
+	 * This avoids the overhead of the mod operation in latency-sensitive cases.
+	 */
+	for (int i = num_stripes; i >= 1; i--) {
```
Comment (bwbarrett) [inline src/nccl_ofi_scheduler.c:53]: As I said before, this should really be `i > 1`, not `i >= 1`.

**Review by bwbarrett — CHANGES_REQUESTED** [review]

File: src/nccl_ofi_scheduler.c (hunk 8 above)
Comment (arunkarthik-akkart) [reply src/nccl_ofi_scheduler.c]: Sorry. I have corrected now.

**Review by bwbarrett — APPROVED** [review]
